### PR TITLE
TRUNK-5169:Fixed the NonUniqueObjectException that breaks subsequent …

### PR DESCRIPTION
**Pull request by samuelmale**
Ticket : https://issues.openmrs.org/browse/TRUNK-5169

### CODE REVIEW

**Comment by samuelmale** [comment]
Just updated my changes here, do you mind looking at it guys.
cc: @dkayiwa @wluyima @djazayeri  etc..

**Comment by samuelmale** [comment]
How was this PR closed guys!
cc @dkayiwa

**Comment by samuelmale** [comment]
Soory, I had forgotten to commit again after squashing my preceding commits, now good 
cc @dkayiwa

**Comment by teleivo** [comment]
@samuelmale can you please update on the latest master branch, to resolve the PR conflicts.

**Comment by samuelmale** [comment]
I have hopefully resolved the merge conflicts @teleivo

**Comment by samuelmale** [comment]
Request for your review @teleivo , @dkayiwa , @wluyima

**Comment by dkayiwa** [comment]
@samuelmale on the JIRA ticket, i put two unit tests. You added only one of them.

**Comment by dkayiwa** [comment]
@samuelmale are you still working on this?

**Comment by samuelmale** [comment]
I made some more changes @dkayiwa

**Comment by dkayiwa** [comment]
The ticket status shows you are still doing more work on this.

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePersonDAO.java [hunk 1]
```diff
@@ -322,10 +324,20 @@ public void deletePersonAttributeType(PersonAttributeType type) {
 	/**
 	 * @see org.openmrs.api.PersonService#savePersonAttributeType(org.openmrs.PersonAttributeType)
 	 * @see org.openmrs.api.db.PersonDAO#savePersonAttributeType(org.openmrs.PersonAttributeType)
+	 * @see org.openmrs.api.util.OpenmrsUtil#isObjectLoadedInCurrentHibernateSession(Session,Class<?>,Id)
 	 */
 	@Override
 	public PersonAttributeType savePersonAttributeType(PersonAttributeType type) {
-		sessionFactory.getCurrentSession().saveOrUpdate(type);
+	    boolean typeAlreadyLoadedInSession = false;
+	    //Check whether type is already loaded in current Hibernate Session
```
Comment (samuelmale) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernatePersonDAO.java]: Hope this comment makes sense @dkayiwa

**Comment by samuelmale** [comment]
Could you look at this one more time @dkayiwa !

**Comment by dkayiwa** [comment]
@samuelmale is this a fix for all openmrs object types as reported in my ticket comments? Or just for some of them?

**Comment by samuelmale** [comment]
It's for all Openmrs objects @dkayiwa

**Comment by dkayiwa** [comment]
Can you add a test for any such objects and prove it works?

**Comment by samuelmale** [comment]
I added a test @dkayiwa

**Comment by dkayiwa** [comment]
What is the name of the test and which openmrs object is it testing?

File: api/src/test/java/org/openmrs/api/PersonServiceTest.java [hunk 2]
```diff
@@ -441,6 +442,38 @@ public void savePersonAttributeType_shouldUpdateAnyGlobalPropertyWhichReferenceT
...
+	
+	@Test
+	public void savePersonAttributeType_shouldNotResultIntoNonUniqueObjectExceptionWithOtherOpenmrsObjects() throws Exception {	
+		PatientIdentifierType type = ps.getPatientIdentifierType(1);
+		Person anotherOpemrsObject = personService.getPerson(1);
+		PersonService service = Context.getPersonService();
+		PersonAttributeType pat = service.getPersonAttributeType(1);
+		service.savePersonAttributeType(pat);
+		try {
+			ps.savePatientIdentifierType(type);
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/api/PersonServiceTest.java]: I thought this is enough prof @dkayiwa

File: api/src/test/java/org/openmrs/api/PersonServiceTest.java [hunk 3]
```diff
@@ -441,6 +442,38 @@ public void savePersonAttributeType_shouldUpdateAnyGlobalPropertyWhichReferenceT
...
+			personService.savePerson(anotherOpemrsObject);
+
```
[the remark this answers is not in the document]
Comment (samuelmale) [reply api/src/test/java/org/openmrs/api/PersonServiceTest.java]: But I have just added this one when you requested me to add one. And fyi this is the second one, you want a 3rd one? @dkayiwa

[the remark this answers is not in the document]
Comment (dkayiwa) [reply api/src/test/java/org/openmrs/api/PersonServiceTest.java]: person and patient or person attribute types are too related. Try something like concept

[the remark this answers is not in the document]
Comment (samuelmale) [reply api/src/test/java/org/openmrs/api/PersonServiceTest.java]: Well the fix I made should always be called explicitly in DAOs. Calling it resolved the issue with `ConceptClass` types.

**Comment by gitcliff** [comment]
@samuelmale  KINDLY look in to the conflicting files

**Comment by samuelmale** [comment]
> @samuelmale KINDLY look in to the conflicting files

@gitcliff I'm actually not actively working on this, someone else is free to pick this up.

**Pull request closed without merging**